app/Http/Controllers/DeviceGroupController.php

```diff
@@ -25,11 +26,34 @@ public function index()
     {
         $this->authorize('manage', DeviceGroup::class);
 
+        $device_groups = DeviceGroup::orderBy('name')->with('devices:devices.device_id')->withCount('devices')->get();
```

Comment:
Instead of eager loading devices.device_id, you could just run a new query with a subselect:
`Device::orderBy('hostname')->doesntHave('groups')->get()`  (or more directly `Device::orderBy('hostname')->whereNotIn('device_id', function ($q) { $q->select('device_id')->from('device_group_device'); })->get()`)

Basically reduces all your changes to a one liner.

Reply:
changed